### PR TITLE
Add handler for Firefox 110

### PR DESCRIPTION
**Pull request by jpgneves**
Adds a handler for Firefox 110 and above where we can invoke `addTransceiver` with the `sendEncodings` parameter.

### CODE REVIEW

**Comment by kompfner** [comment]
Is this meant for daily-web or daily-js? If the former, shouldn't this PR be against the `v3-daily` branch? And if the latter, I think we need to turn this into a patch that's applied at build time, since `daily-js` doesn't use our fork of `mediasoup-client` (there are currently multiple patches applied this way).

File: src/handlers/Firefox110.ts [hunk 1]
```diff
@@ -0,0 +1,1041 @@
...
+import { RemoteSdp } from './sdp/RemoteSdp';
+import { IceParameters, DtlsRole } from '../Transport';
+import { RtpCapabilities, RtpParameters } from '../RtpParameters';
+import { SctpCapabilities, SctpStreamParameters } from '../SctpParameters';
+
+const logger = new Logger('Firefox110');
+
+const SCTP_NUM_STREAMS = { OS: 16, MIS: 2048 };
+
+export class Firefox110 extends HandlerInterface
```
Comment (kompfner) [inline src/handlers/Firefox110.ts:30]: Is this class a wholesale copy of a Chrome handler now? (And if so, which one?)

Comment (kompfner) [reply src/handlers/Firefox110.ts:30]: Huh, didn't realize they added a Chrome 111 one...that must've been pretty recent.

**Comment by kompfner** [comment]
> Is this meant for daily-web or daily-js? If the former, shouldn't this PR be against the `v3-daily` branch? And if the latter, I think we need to turn this into a patch that's applied at build time, since `daily-js` doesn't use our fork of `mediasoup-client` (there are currently multiple patches applied this way).

Oh, or is this in preparation for opening a PR against upstream? Is the plan to simply try to get it fixed at the source and not bother with also patching our own code in the meantime, since it's just a relatively harmless warning that we're contending with?

File: src/handlers/Firefox110.ts (hunk 1 above)
Comment (jpgneves) [reply src/handlers/Firefox110.ts:30]: It's a copy of the Firefox60 handler, but tweaked to be more in line with what is now supported in Firefox 110 (namely `getNativeRtpCapabilities` is now based on the Chrome implementation and `send` now includes `sendEncodings` as a parameter to `addTransceiver`).

**Comment by jpgneves** [comment]
> > Is this meant for daily-web or daily-js? If the former, shouldn't this PR be against the `v3-daily` branch? And if the latter, I think we need to turn this into a patch that's applied at build time, since `daily-js` doesn't use our fork of `mediasoup-client` (there are currently multiple patches applied this way).
> 
> Oh, or is this in preparation for opening a PR against upstream? Is the plan to simply try to get it fixed at the source and not bother with also patching our own code in the meantime, since it's just a relatively harmless warning that we're contending with?

Yeah, ideally we can just get this upstreamed - it affects both `daily-js` and `daily-web` at the moment. The warning will turn into an error at some point, and would be nice to not have them popping up in any case. This still won't resolve _all_ of them (because there's an actual Firefox bug which is getting fixed in the next release), but it will fix the deprecated API usage.

**Comment by filipi87** [comment]
Hi @jpgneves , and as @kompfner had mentioned before, is It already worth turn this into a patch that's applied at build time for `daily-js`, while these fixes are not included on `mediasoup-client` upstream ?

File: src/handlers/Firefox110.ts [hunk 2]
```diff
@@ -0,0 +1,1041 @@
...
+			sendingRtpParameters.encodings.length > 1 &&
+			(
+				sendingRtpParameters.codecs[0].mimeType.toLowerCase() === 'video/vp8' ||
+				sendingRtpParameters.codecs[0].mimeType.toLowerCase() === 'video/h264'
+			)
+		)
+		{
+			for (const encoding of sendingRtpParameters.encodings)
+			{
+				encoding.scalabilityMode = 'S1T3';
```
Comment (filipi87) [inline src/handlers/Firefox110.ts:395]: In all the other handlers, `Firefox60`, `Chrome74` or `Chrome111`, we retrieve this number from the `temporalLayers`, shouldn't we do the same here ?

File: src/handlers/Firefox110.ts [hunk 3]
```diff
@@ -0,0 +1,1041 @@
...
+
+		parameters.encodings.forEach((encoding: RTCRtpEncodingParameters, idx: number) =>
+		{
+			if (idx >= spatialLayer)
+				encoding.active = true;
+			else
+				encoding.active = false;
+		});
+
+		await transceiver.sender.setParameters(parameters);
```
Comment (filipi87) [inline src/handlers/Firefox110.ts:586]: I believe after this line, we should do what mediasoup has done in this [commit](https://github.com/versatica/mediasoup-client/commit/9000832d3b2193925c9492882a7a0ef230bbdc86), to persist the changes in sending encoding:

```suggestion
		await transceiver.sender.setParameters(parameters);

		this._remoteSdp!.muxMediaSectionSimulcast(localId, parameters.encodings);

		const offer = await this._pc.createOffer();

		logger.debug(
			'setMaxSpatialLayer() | calling pc.setLocalDescription() [offer:%o]',
			offer);

		await this._pc.setLocalDescription(offer);

		const answer = { type: 'answer', sdp: this._remoteSdp!.getSdp() };

		logger.debug(
			'setMaxSpatialLayer() | calling pc.setRemoteDescription() [answer:%o]',
			answer);

		await this._pc.setRemoteDescription(answer);
```

Comment (filipi87) [reply src/handlers/Firefox110.ts:586]: We have already faced an issue with Chrome in the past, where after we start to screen share or send a custom track, the bandwidth was not respected anymore. 

And these changes from mediasoup fix this kind of behavior.

File: src/handlers/Firefox110.ts (hunk 2 above)
Comment (filipi87) [reply src/handlers/Firefox110.ts:395]: We have also a recent commit from mediasoup related to this: 
[Fix scalabilityMode: S1T2 is not valid for VP8, use L1T2 instead](https://github.com/versatica/mediasoup-client/pull/247)

**Comment by aconchillo** [comment]
Closing this. We have Firefox117 and upstream has Firefox120.

**Pull request closed without merging**